### ISSUE TITLE
Suggestion: Publish packages directly to ~/.pnpm-store

### ISSUE THREAD

**Issue by KSXGitHub** [pnpm/pnpm#2053]
It is annoying to download my own packages in order to use them in other repos. So in order to save time and bandwidth, I suggest pnpm adds ability to make copies of my packages in `~/.pnpm-store` whenever I type `pnpm publish`.

## Suggestions

* `pnpm publish` uploads files to the npm registry first (1), then copies them to `~/.pnpm-store` (2).

* By default, both (1) and (2) is enabled.

* `--no-local-publish` flag (or `local-publish = false` in `.npmrc`) disables (2) preventing pnpm from making local copies.

* `--local-publish-only` flag (or `local-publish-only = true` in `.npmrc`) disables (1) preventing pnpm from uploading files to the registry (only makes local copy).

* If pnpm fails to upload files to the registry, it should not make local copies.

* If pnpm successes in uploading files to the registry, it should always update local copies.

**Comment by suricactus** [comment]
I came to the issue tracker to request exactly the same feature. Currently we're using custom script to include the packages within the whole store and metadata storage of pnpm, but on the new major version our custom script do not longer work. 
Can someone from the contributors give an approximation of how much work it will require?

**Comment by zkochan** [comment]
@suricactus I am not sure I understand your case. You commit the pnpm store to the repository?

**Comment by zkochan** [comment]
I think this would need a separate issue. 

Adding support for adding tarballs to the store is probably not hard. (I am surprised that `pnpm store add file:./path.tgz` doesn't work already.

However, local tarballs are not saved at the same location where regular dependencies. So even if you add the local foo tarball to the store, `pnpm add foo` will not use the tarball you added.

I have to suggestions how you can do this w/o adding changes to pnpm

1. use pnpm's [workspace feature](https://pnpm.js.org/en/workspaces)
1. move all your packages to a scope. Run a [Verdaccio](https://verdaccio.org/) server locally or on a server and publish your packages to Verdaccio.

**Comment by suricactus** [comment]
Thanks!
It is good to know that tarballs are stored in a different manner. I already use a different "registry" in my script to incorporate the tarball there.

It's already a monorepo and all the packages are scoped, however, the situation is a bit more complicated. In my case the pnpm store is used for package distribution and  pnpm stays completely offline on production installations. However, there are other packages outside the monorepo that require older versions of some of the monorepo packages. Except adding a new package without using a online registry. That's why I had a script in pnpm@3, that mimics the installation process from a registry. At that stage we considered it would be easier to have our custom version, but as breaking changes are introduced in pnpm@4, I would be happy to see this feature natively supported.
 
I've also considered the Verdaccio, but it seems to be overkill. For a simple task like publishing a package directly into the pnpm store.

**Comment by whitecolor** [comment]
There is a tool for using with WIP packages locally https://github.com/whitecolor/yalc

**Comment by thobson** [comment]
I'd like something similar but for a different use case:

I have two packages in a workspace

* @passlock/shared
* @passlock/client  - references "@passlock/shared": "workspace:*"

I want to test @passlock/client from outside the workspace e.g.

* @passlock/client  - references  "@passlock/client": "file:../../Passlock/packages/client"

But I get an error:

In : "@passlock/shared@workspace:*" is in the dependencies but no package named "@passlock/shared" is present in the workspace

Basically I want to check that everything hangs together ok before publishing to npmjs, but so far I haven't found a way to do this, other than moving everything into a single workspace. This isn't ideal because it's not representative of how other developers would use @passlock/client.

Of course I'm open to any other suggestions or workarounds.
